chore: prepare `Cargo.toml`, `README.md`, `DEVELOPMENT.md` ready for initial release (#48)
CQCL/hugr-llvm: DEVELOPMENT.md

# Welcome to the hugr-llvm development guide <!-- omit in toc -->

This guide is intended to help you get started with developing hugr-llvm.

If you find any errors or omissions in this document, please [open an issue](https://github.com/CQCL/hugr-llvm/issues/new)!

## #️⃣ Setting up the development environment

> You can setup the development environment in two ways:

Reviewer: ```suggestion
You can set up the development environment in two ways:
```

### The Nix way

> The easiest way to setup the development environment is to use the provided

Reviewer: ```suggestion
The easiest way to set up the development environment is to use the provided
```

> [`devenv.nix`](devenv.nix) file. This will setup a development shell with all the

Reviewer: ```suggestion
[`devenv.nix`](devenv.nix) file. This will set up a development shell with all the
```

required dependencies.

To use this, you will need to install [devenv](https://devenv.sh/getting-started/).
Once you have it running, open a shell with:

```bash
devenv shell
```

All the required dependencies should be available. You can automate loading the
shell by setting up [direnv](https://devenv.sh/automatic-shell-activation/).

### Manual setup

> To setup the environment manually you will need:

Reviewer: ```suggestion
To set up the environment manually you will need:
```

- Rust `>=1.75`: https://www.rust-lang.org/tools/install
- llvm `== 14.0`: we use the rust bindings 
[llvm-sys](https://crates.io/crates/llvm-sys) to [llvm](https://llvm.org/), 
- Poetry `>=1.8`: https://python-poetry.org/

Once you have these installed, verify that your setup is working

```bash
cargo test
```

## 💅 Coding Style

We use `rustfmt` to enforce a consistent coding style. The CI will fail if the code is not formatted correctly.

To format your code, run:

```bash
cargo format
```

> We also use various linters to catch common mistakes and enforce best practices. To run these, see [out ci config](./.github/workflows/ci-rs.yml). TODO Provide a better way, contributions welcome.

Reviewer: ```suggestion
We also use various linters to catch common mistakes and enforce best practices. To run these, see [our CI config](./.github/workflows/ci-rs.yml). TODO Provide a better way, contributions welcome.
```

## 📈 Code Coverage

We run coverage checks on the CI. Once you submit a PR, you can review the
line-by-line coverage report on
[codecov](https://app.codecov.io/gh/CQCL/hugr-llvm/commits?branch=All%20branches).

> To run the coverage checks locally, first install `cargo-llvm-cov`.
> ```bash

Reviewer: ```suggestion
To run the coverage checks locally, first install `cargo-llvm-cov`.

```bash
```

cargo install cargo-llvm-cov
```

> Then run the tests, see [our CI config](/.github/workflows/ci-rs.yml).

Reviewer: ```suggestion
Then run the tests:
```
?

```

This will generate a coverage file that can be opened with your favourite coverage viewer. In VSCode, you can use
[`coverage-gutters`](https://marketplace.visualstudio.com/items?itemName=ryanluker.vscode-coverage-gutters).

## 🌐 Contributing to hugr-llvm

We welcome contributions to hugr-llvm! Please open [an issue](https://github.com/CQCL/hugr-llvm/issues/new) or [pull request](https://github.com/CQCL/hugr-llvm/compare) if you have any questions or suggestions.

PRs should be made against the `main` branch, and should pass all CI checks before being merged. This includes using the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) format for the PR title.

Some tests may be skipped based on the changes made. To run all the tests in
your PR mark it with a 'run-ci-checks' label and push new commits to it.

The general format of a contribution title should be:

```
<type>(<scope>)!: <description>
```

Where the scope is optional, and the `!` is only included if this is a semver breaking change that requires a major version bump.

We accept the following contribution types:

- feat: New features.
- fix: Bug fixes.
- docs: Improvements to the documentation.

> - style: Formatting, missing semi colons, etc; no code change.

Reviewer: ```suggestion
- style: Formatting, missing semicolons, etc; no code change.
```

- refactor: Refactoring code without changing behaviour.
- perf: Code refactoring focused on improving performance.
- test: Adding missing tests, refactoring tests; no production code change.

> - ci: CI related changes. These changes are not published in the changelog.

Reviewer: ```suggestion
- ci: CI-related changes. These changes are not published in the changelog.
```
